fem/gslib/findpts_local_2.cpp

```diff
@@ -0,0 +1,1246 @@
+                                    const int j,
+                                    const int pN)
+{
+   if (j >= pN)
+   {
+      return;
+   }
+
+   dist2[j] = DBL_MAX;
+
+   double zr = z[j];
+   for (int k = 0; k < pN; ++k)
+   {
+      double zs = z[k];
+      const int jk = j + k * pN;
+      double dx[2];
+      for (int d = 0; d < 2; ++d)
+      {
+         dx[d] = x[d] - elx[d][jk];
+      }
+      const double dist2_jkl = l2norm2(dx);
+      if (dist2[j] > dist2_jkl)
+      {
+         dist2[j] = dist2_jkl;
+         r[0][j] = zr;
+         r[1][j] = zs;
+      }
+   }
+}
+
+/* Compute contribution towards function value and its derivatives in each
+   reference direction. */
+static MFEM_HOST_DEVICE double tensor_ig2_j(double *g_partials,
+                                            const double *Jr,
+                                            const double *Dr,
+                                            const double *Js,
+                                            const double *Ds,
+                                            const double *u,
+                                            const int j,
+                                            const int pN)
+{
+   double uJs = 0.0;
+   double uDs = 0.0;
+   for (int k = 0; k < pN; ++k)
+   {
+      uJs += u[j + k * pN] * Js[k];
+      uDs += u[j + k * pN] * Ds[k];
+   }
+
+   g_partials[0] = uJs * Dr[j];
+   g_partials[1] = uDs * Jr[j];
+   return uJs * Jr[j];
+}
+
+template<int T_D1D = 0>
+static void FindPointsLocal2D_Kernel(const int npt,
+                                     const double tol,
+                                     const double *x,
+                                     const int point_pos_ordering,
+                                     const double *xElemCoord,
+                                     const int nel,
+                                     const double *wtend,
+                                     const double *boxinfo,
+                                     const int hash_n,
+                                     const double *hashMin,
+                                     const double *hashFac,
+                                     unsigned int *hashOffset,
+                                     unsigned int *const code_base,
+                                     unsigned int *const el_base,
+                                     double *const r_base,
+                                     double *const dist2_base,
+                                     const double *gll1D,
+                                     const double *lagcoeff,
+                                     const int pN = 0)
+{
+#define MAX_CONST(a, b) (((a) > (b)) ? (a) : (b))
+   const int MD1 = T_D1D ? T_D1D : 14;
+   const int D1D = T_D1D ? T_D1D : pN;
+   const int p_NE = D1D*D1D;
+   const int p_NEL = nel*p_NE;
+   MFEM_VERIFY(MD1 <= 14,"Increase Max allowable polynomial order.");
+   MFEM_VERIFY(D1D != 0, "Polynomial order not specified.");
+   const int nThreads = 32;
```

Comment:
Should this be a tune-able parameter?

Reply:
yeah I could define it as `MD1*DIM` inside the kernel?